x-pack/filebeat/input/netflow/decoder/v9/lru.go

```diff
@@ -0,0 +1,164 @@
+// Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+// or more contributor license agreements. Licensed under the Elastic License;
+// you may not use this file except in compliance with the Elastic License.
+
+package v9
+
+import (
+	"bytes"
+	"container/heap"
+	"sync"
+	"sync/atomic"
+	"time"
+)
+
+type eventWithMissingTemplate struct {
+	key       SessionKey
+	entryTime time.Time
+}
+
+type pendingEventsHeap []eventWithMissingTemplate
+
+func (h pendingEventsHeap) Len() int {
+	return len(h)
+}
+
+func (h pendingEventsHeap) Less(i, j int) bool {
+	return h[i].entryTime.Sub(h[j].entryTime) < 0
+}
+
+func (h pendingEventsHeap) Swap(i, j int) {
+	h[i], h[j] = h[j], h[i]
+}
+
+func (h *pendingEventsHeap) Push(x any) {
+	v, ok := x.(eventWithMissingTemplate)
+	if ok {
+		*h = append(*h, v)
+	}
+}
+
+func (h *pendingEventsHeap) Pop() any {
+	old := *h
+	n := len(old)
+	x := old[n-1]
+	*h = old[0 : n-1 : n-1]
+	return x
+}
+
+type pendingTemplatesCache struct {
+	mtx     sync.RWMutex
+	wg      sync.WaitGroup
+	isEmpty atomic.Bool
+	hp      pendingEventsHeap
+	started bool
+	events  map[SessionKey][]*bytes.Buffer
+}
+
+func newPendingTemplatesCache() *pendingTemplatesCache {
+	cache := &pendingTemplatesCache{
+		events: make(map[SessionKey][]*bytes.Buffer),
+		hp:     pendingEventsHeap{},
+	}
+	return cache
+}
+
+// GetAndRemove returns all events for a given session key and removes them from the cache
+func (h *pendingTemplatesCache) GetAndRemove(key SessionKey) []*bytes.Buffer {
+	if h.isEmpty.Load() {
+		return nil
+	}
+
+	h.mtx.Lock()
+	defer h.mtx.Unlock()
+	events, ok := h.events[key]
+	if !ok {
+		return nil
+	}
+	delete(h.events, key)
+	h.isEmpty.Store(len(h.events) == 0)
+	return events
+}
+
+// Add adds an event to the pending templates cache
+func (h *pendingTemplatesCache) Add(key SessionKey, events *bytes.Buffer) {
+	h.mtx.Lock()
+	defer h.mtx.Unlock()
+
+	h.events[key] = append(h.events[key], events)
+	h.hp.Push(eventWithMissingTemplate{key: key, entryTime: time.Now()})
+	h.isEmpty.Store(false)
```

Comment:
The use of isEmpty has race condition, for example:
1. first go routine passes line 89, pushed the data in, the isEmpty is still true though
2. second go routine Loads isEmpty on line 68 that returns true, thinks that it's still empty, misses the data that was just added
3. first go routine marks IsEmpty false on line 90

Remove isEmpty atomic completely, synchronize access to len(h.events)

Reply:
thanks for catching that, I followed your suggestion `isEmpty` removed. Feel free to unresolve if you spot anything else